commons/src/test/java/io/aiven/kafka/tiered/storage/commons/transform/TransformsEndToEndTest.java

```diff
@@ -0,0 +1,107 @@
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TransformsEndToEndTest extends AesKeyAwareTest {
+    static final int ORIGINAL_SIZE = 1812004;
+
+    static byte[] original;
+
+    @BeforeAll
+    static void init() {
+        original = new byte[ORIGINAL_SIZE];
+        final var random = new Random();
+        random.nextBytes(original);
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 2, 3, 5, 13, 1024, 1024 * 2, 1024 * 5 + 3, ORIGINAL_SIZE - 1, ORIGINAL_SIZE * 2})
+    void plaintext(final int chunkSize) throws IOException {
+        test(chunkSize, false, false);
+    }
+
+    @ParameterizedTest
+    // Small chunks would make encryption and compression tests very slow, skipping them
+    @ValueSource(ints = {1024 - 1, 1024, 1024 * 2 + 2, 1024 * 5 + 3, ORIGINAL_SIZE - 1, ORIGINAL_SIZE * 2})
+    void encryption(final int chunkSize) throws IOException {
+        test(chunkSize, false, true);
+    }
+
+    @ParameterizedTest
+    // Small chunks would make compression tests going very slowly, skipping them
+    @ValueSource(ints = {1024 - 1, 1024, 1024 * 2 + 2, 1024 * 5 + 3, ORIGINAL_SIZE - 1, ORIGINAL_SIZE * 2})
+    void compression(final int chunkSize) throws IOException {
+        test(chunkSize, true, false);
+    }
+
+    @ParameterizedTest
+    // Small chunks would make compression tests going very slowly, skipping them
+    @ValueSource(ints = {1024 - 1, 1024, 1024 * 2 + 2, 1024 * 5 + 3, ORIGINAL_SIZE - 1, ORIGINAL_SIZE * 2})
+    void compressionAndEncryption(final int chunkSize) throws IOException {
+        test(chunkSize, true, true);
+    }
+
+    private void test(final int chunkSize, final boolean compression, final boolean encryption) throws IOException {
+        // Transform.
+        TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(
+            new ByteArrayInputStream(original), chunkSize);
+        if (compression) {
+            transformEnum = new CompressionChunkEnumeration(transformEnum);
+        }
+        if (encryption) {
+            transformEnum = new EncryptionChunkEnumeration(transformEnum, AesKeyAwareTest::encryptionCipherSupplier);
+        }
+        final var transformFinisher = new TransformFinisher(transformEnum, ORIGINAL_SIZE);
+        final byte[] uploadedData;
+        final ChunkIndex chunkIndex;
+        try (final var sis = new SequenceInputStream(transformFinisher)) {
+            uploadedData = sis.readAllBytes();
+            chunkIndex = transformFinisher.chunkIndex();
+        }
+
+        // Detransform.
+        DetransformChunkEnumeration detransformEnum = new BaseDetransformChunkEnumeration(
+            new ByteArrayInputStream(uploadedData), chunkIndex.chunks());
+        if (encryption) {
+            assert secretKey != null;
```

Comment:
Should this be `assertThat(secretKey).isNotNull()` with a proper message explaining the failure?

Reply:
This is not really a part of the test, just a sanity check for the test code itself. But yeah, we can do this, why not

Reply:
I understand that this is a sanity check but anyways `assert` I believe should not be used in tests and also can be disabled by JVM flags